File: archive/cdktf-py/Pr4930/lib/cache_stack.py
"""ElastiCache stack for caching layer."""

from constructs import Construct
from cdktf_cdktf_provider_aws.elasticache_subnet_group import ElasticacheSubnetGroup
from cdktf_cdktf_provider_aws.elasticache_replication_group import ElasticacheReplicationGroup


class CacheStack(Construct):
    """ElastiCache cluster for product data caching."""

    def __init__(
        self,
        scope: Construct,
        construct_id: str,
        *,
        vpc_id: str,
        private_subnet_ids: list,
        cache_security_group_id: str,
        environment_suffix: str
    ):
        """Initialize ElastiCache cluster."""
        super().__init__(scope, construct_id)

        # Create subnet group for ElastiCache
        subnet_group = ElasticacheSubnetGroup(
            self,
            "cache_subnet_group",
            name=f"pc-cache-subnet-{environment_suffix}",
            description="Subnet group for product catalog cache",
            subnet_ids=private_subnet_ids,
            tags={
                "Name": f"product-catalog-cache-subnet-{environment_suffix}"
            }
        )

        # Create ElastiCache replication group (Valkey/Redis)
        self.cache_cluster = ElasticacheReplicationGroup(
            self,
            "cache_cluster",
            replication_group_id=f"pc-{environment_suffix}",
            description="Product catalog cache cluster",
            engine="valkey",
            engine_version="8.0",
            node_type="cache.t3.micro",
            num_cache_clusters=1,
            port=6379,
            subnet_group_name=subnet_group.name,
            security_group_ids=[cache_security_group_id],
            tags={
                "Name": f"product-catalog-cache-{environment_suffix}"
            }
        )

    @property
    def cache_endpoint(self):
        """Return cache primary endpoint."""
        return self.cache_cluster.primary_endpoint_address

File: archive/cdk-ts/Pr4248/lib/cache-stack.ts
import * as cdk from 'aws-cdk-lib';
import * as ec2 from 'aws-cdk-lib/aws-ec2';
import * as elasticache from 'aws-cdk-lib/aws-elasticache';
import * as kms from 'aws-cdk-lib/aws-kms';
import * as logs from 'aws-cdk-lib/aws-logs';
import { Construct } from 'constructs';

interface CacheStackProps {
  environmentSuffix: string;
  vpc: ec2.Vpc;
  cacheSecurityGroup: ec2.SecurityGroup;
  kmsKey: kms.Key;
}

export class CacheStack extends Construct {
  public readonly redisEndpoint: string;

  constructor(scope: Construct, id: string, props: CacheStackProps) {
    super(scope, id);

    // Create CloudWatch log group for ElastiCache logs
    const logGroup = new logs.LogGroup(this, 'RedisLogGroup', {
      logGroupName: `/aws/elasticache/redis/${props.environmentSuffix}`,
      retention: logs.RetentionDays.ONE_WEEK,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
    });

    // Create ElastiCache subnet group
    const subnetGroup = new elasticache.CfnSubnetGroup(
      this,
      'CacheSubnetGroup',
      {
        description: 'Subnet group for ElastiCache Redis',
        subnetIds: props.vpc.privateSubnets.map(subnet => subnet.subnetId),
        cacheSubnetGroupName: `payment-cache-subnet-${props.environmentSuffix}`,
      }
    );

    // Create ElastiCache replication group (Redis cluster)
    const replicationGroup = new elasticache.CfnReplicationGroup(
      this,
      'RedisCluster',
      {
        replicationGroupId: `payment-redis-${props.environmentSuffix}`,
        replicationGroupDescription: 'Redis cluster for payment processing',
        engine: 'redis',
        engineVersion: '7.1',
        cacheNodeType: 'cache.t3.micro',
        numCacheClusters: 1,
        multiAzEnabled: false,
        automaticFailoverEnabled: false,
        autoMinorVersionUpgrade: true,
        cacheSubnetGroupName: subnetGroup.cacheSubnetGroupName,
        securityGroupIds: [props.cacheSecurityGroup.securityGroupId],
        atRestEncryptionEnabled: true,
        transitEncryptionEnabled: true,
        transitEncryptionMode: 'required',
        kmsKeyId: props.kmsKey.keyId,
        snapshotRetentionLimit: 7,
        snapshotWindow: '03:00-05:00',
        preferredMaintenanceWindow: 'sun:05:00-sun:07:00',
        logDeliveryConfigurations: [
          {
            destinationType: 'cloudwatch-logs',
            logFormat: 'json',
            logType: 'slow-log',
            destinationDetails: {
              cloudWatchLogsDetails: {
                logGroup: logGroup.logGroupName,
              },
            },
          },
        ],
        tags: [
          { key: 'PCICompliant', value: 'true' },
          { key: 'Environment', value: props.environmentSuffix },
        ],
      }
    );

    replicationGroup.addDependency(subnetGroup);
    // Ensure log group is created before replication group
    replicationGroup.addDependency(
      logGroup.node.defaultChild as cdk.CfnResource
    );

    // Store endpoint for reference
    this.redisEndpoint = replicationGroup.attrPrimaryEndPointAddress;

    // Output endpoint
    new cdk.CfnOutput(cdk.Stack.of(this), 'RedisClusterEndpoint', {
      value: this.redisEndpoint,
      description: 'Redis cluster primary endpoint',
    });
  }
}
